### PR TITLE
fix(*): migrate rollup to tsup, support esm correctly, remove *.spec.* for dist

### LINKED ISSUE #19: Excluding *.spec.ts files is required for rollup config
https://www.npmjs.com/package/es-toolkit?activeTab=code

image

### LINKED ISSUE #20: Maybe we might try just using tsup's declaration generation...
Maybe we might try just using tsup's declaration generation...

_Originally posted by @raon0211 in https://github.com/toss/es-toolkit/pull/16#discussion_r1624466058_

### PR DESCRIPTION
**Pull request by manudeli**
close #19 close #20 

@raon0211 In my opinion, This change is fit for your requirement. Review strictly please!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,49 +17,83 @@
     "./package.json": "./package.json"
   },
   "files": [
-    "dist/**/*",
-    "esm/**/*",
-    "*.d.ts"
```
Comment (manudeli) [inline package.json:22]: tsup's default use only dist

File: package.json [hunk 2]
```diff
@@ -17,49 +17,83 @@
...
+    "dist"
   ],
   "publishConfig": {
     "access": "public",
     "main": "./dist/index.js",
+    "module": "./dist/index.mjs",
```
Comment (manudeli) [inline package.json:25]: module field for yarn berry
https://yarnpkg.com/configuration/manifest#module

File: package.json [hunk 3]
```diff
@@ -17,49 +17,83 @@
...
     "types": "./dist/index.d.ts",
     "exports": {
       ".": {
-        "types": "./dist/index.d.ts",
-        "import": "./esm/index.mjs",
-        "require": "./dist/index.js"
+        "import": {
+          "types": "./dist/index.d.mts",
```
Comment (manudeli) [inline package.json:30]: *.d.mts is generated by tsup default

https://github.com/toss/es-toolkit/pull/16#discussion_r1624466058

File: tsup.config.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { defineConfig } from 'tsup'
+
+export default defineConfig({
+  format: ['cjs', 'esm'],
+  entry: ['src/*.ts', 'src/*/*.ts', '!**/*.{spec,test,test-d}.*'],
```
Comment (manudeli) [inline tsup.config.ts:5]: `!**/*.{spec,test,test-d}.*`: This prevent that dist contain test files

File: package.json [hunk 5]
```diff
@@ -84,14 +117,14 @@
...
-    "rollup": "^2.78.0",
+    "tsup": "^8.1.0",
     "typescript": "^5.4.5",
     "vitest": "^1.5.2"
   },
   "sideEffects": false,
   "scripts": {
     "prepack": "yarn build",
-    "build": "rm -rf dist esm && tsc -p tsconfig.json --declaration --emitDeclarationOnly --declarationDir dist && rollup -c rollup.config.js && ./.scripts/postbuild.sh",
+    "build": "tsup",
```
Comment (manudeli) [inline package.json:127]: So simple

**Review by manudeli — COMMENTED** [review]
Comments to share intention

**Comment by raon0211** [comment]
image

Fixed the Node10 resolution in the commit in https://github.com/toss/es-toolkit/pull/21/commits/755b938d391cbfdfd86473f11e1a86c4c96549e1.

File: .scripts/postbuild.sh [hunk 6]
```diff
@@ -3,4 +3,4 @@ echo "export * from './dist/function';" > function.d.ts
 echo "export * from './dist/math';" > math.d.ts
 echo "export * from './dist/object';" > object.d.ts
 echo "export * from './dist/predicate';" > predicate.d.ts
-echo "export * from './dist/promise';" > promise.d.ts
+echo "export * from './dist/promise';" > promise.d.ts
```
Comment (raon0211) [inline .scripts/postbuild.sh:6]:
```suggestion
echo "export * from './dist/promise';" > promise.d.ts

```

**Review by raon0211 — DISMISSED** [review]
Thanks for your contribution! I wish we had time to better configure the settings to generate declaration maps and make the chunk names reflect their actual names. (rather than `chunk-*.mjs` files..)

File: package.json [hunk 7]
```diff
@@ -84,14 +118,14 @@
...
     "vitest": "^1.5.2"
   },
   "sideEffects": false,
   "scripts": {
     "prepack": "yarn build",
-    "build": "rm -rf dist esm && tsc -p tsconfig.json --declaration --emitDeclarationOnly --declarationDir dist && rollup -c rollup.config.js && ./.scripts/postbuild.sh",
+    "build": "tsup && ./.scripts/postbuild.sh",
     "test": "vitest run --coverage --typecheck"
   }
 }
```
Comment (manudeli) [inline package.json:131]:
```suggestion
}

```

**Review by raon0211 — APPROVED** [review]

**Pull request merged**